pkg/controllers/metrics/node/controller.go

```diff
@@ -0,0 +1,122 @@
+	ctx = logging.WithLogger(ctx, logging.FromContext(ctx).Named(controllerName))
+
+	provisionerName := req.NamespacedName.Name
+
+	// 1. Has the provisioner been deleted?
+	if err := c.provisionerExists(ctx, req); err != nil {
+		if !errors.IsNotFound(err) {
+			// Unable to determine existence of the provisioner, try again later.
+			return reconcile.Result{Requeue: true}, err
+		}
+
+		// The provisioner has been deleted. Reset all the associated counts to zero.
+		if err := publishNodeCountsForProvisioner(provisionerName, consumeNoNodes); err != nil {
+			// One or more metrics were not zeroed. Try again later.
+			return reconcile.Result{Requeue: true}, err
+		}
+
+		// Since the provisioner is gone, do not requeue.
+		return reconcile.Result{}, nil
+	}
+
+	// 2. Update node counts associated with this provisioner.
+	if err := publishNodeCountsForProvisioner(provisionerName, c.consumeNodesFromKubeClientWithContext(ctx)); err != nil {
+		// An updated value for one or more metrics was not published. Try again later.
+		return reconcile.Result{Requeue: true}, err
+	}
+
+	// 3. Schedule the next run.
+	return reconcile.Result{RequeueAfter: requeueInterval}, nil
+}
+
+func (c *Controller) Register(_ context.Context, m manager.Manager) error {
+	return controllerruntime.
+		NewControllerManagedBy(m).
+		Named(controllerName).
+		For(&karpenterv1.Provisioner{}, builder.WithPredicates(
+			predicate.Funcs{
+				CreateFunc:  func(_ event.CreateEvent) bool { return true },
+				DeleteFunc:  func(_ event.DeleteEvent) bool { return true },
+				UpdateFunc:  func(_ event.UpdateEvent) bool { return false },
+				GenericFunc: func(_ event.GenericEvent) bool { return false },
+			},
+		)).
+		WithOptions(controller.Options{
+			MaxConcurrentReconciles: 1,
+		}).
+		Complete(c)
+}
+
+// provisionerExists simply attempts to retrieve the provisioner from the Controller's Client
+// and returns any resulting error.
+func (c *Controller) provisionerExists(ctx context.Context, req reconcile.Request) error {
+	provisioner := karpenterv1.Provisioner{}
+	return c.KubeClient.Get(ctx, req.NamespacedName, &provisioner)
+}
+
+// consumeNodesFromKubeClientWithContext will retrieve matching nodes from the Controller's Client then
+// pass the nodes to `consume` and returns any resulting error. If Client returns an error when
+// retrieving nodes then the error is returned without calling `consume`.
+func (c *Controller) consumeNodesFromKubeClientWithContext(ctx context.Context) consumeNodesWithFunc {
+	return func(nodeLabels client.MatchingLabels, consume nodeListConsumerFunc) error {
+		nodes := corev1.NodeList{}
+		if err := c.KubeClient.List(ctx, &nodes, nodeLabels); err != nil {
+			return err
+		}
+		return consume(nodes.Items)
+	}
+}
+
+// consumeNoNodes calls `consume` with an empty slice and returns any resulting error.
+func consumeNoNodes(_ client.MatchingLabels, consume nodeListConsumerFunc) error {
```

Comment:
I'd suggest a name like `resetNodeCounts` or something like that? (I.e., more of an idea of what it's used for, rather than how it's implemented).

Reply:
The function itself does not reset the counts.  How about `consumeZeroNodes`?

Reply:
Sounds good, thanks